### ISSUE TITLE
drop SinkDocument in PostProcessor

### ISSUE THREAD

**Issue by CGaul** [hpgrahsl/kafka-connect-mongodb#31]
Hey there,
first thank you for this kafka mongodb connector, the code looks well understandable.

I have a question about the PostProcessor interface. I want to implement a PostProcessor, that drops the SinkDocument, when the process(..) internals are failing. What's the best way to implement this? In the MongoDBSinkTask I see that the valueDoc is only added to the docsToWrite List, if there is a valueDoc present. I think what I could do now is to define a getNext in the PostProcessor, like this:
```
getNext().ifPresent(pp -> {
                BsonDocument keyDoc = doc.getKeyDoc().isPresent() ? doc.getKeyDoc().get().clone() : null;
                pp.process(new SinkDocument(keyDoc, null), orig);
            });
```
In order to create a null valueDoc for the next PostProcessor in the Chain. From my understanding, this does only work however, if I would put another PostProcessor at the end of the chain, am I right? So e.g. I would put the DocumentIdAdder at the end, to make sure that the valueDoc is null, if the predecessing PostProcessor (my implementation) failed.

Wouldn't it make sense to change the PostProcessor Interface to return an ``Optional<SinkDocument>``, rather than leave it like ``public void process(SinkDocument doc, SinkRecord orig)``. Then, `` MongoDbSinkTask.buildWriteModel(..)`` would need to evaluate the return value, before the next processor receives the evaluated SinkDocument. In this case you might even not need to use getNext() in every PostProcessor implementation anymore.

This is more a question, rather than an issue right now. I would be happy if you share your thoughts about this.

Best regards,
Constantin

**Comment by CGaul** [comment]
Hi @hpgrahsl,

thanks for the fast reply!
The process that I'm solving in a PostProcessor is the transformation of a String field to a BsonDateTime. In order to archieve that, I'm parsing the field (which needs to be properly formatted ISO-8601 String) and overwrite this field in the valueDoc, like shown in the following snippet:
```
doc.getValueDoc().ifPresent(valueDoc -> {
            if (valueDoc.isString(dateFieldName)) {
                BsonString dateFieldString = valueDoc.getString(dateFieldName);
                try {
                    TemporalAccessor parsedDateTime = DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(dateFieldString.getValue());
                    Instant dateTimeInstant = ZonedDateTime.from(parsedDateTime).toInstant();

                    valueDoc.put(dateFieldName, new BsonDateTime(dateTimeInstant.toEpochMilli()));

                    getNext().ifPresent(pp -> pp.process(doc, orig));

                } catch (DateTimeParseException e) {
                    LOGGER.error("A DateTimeParse exception occurred, as the value of the \"" + dateFieldName + "\""+
                            " field was not parseable from the DateTimeFormatter.ISO_OFFSET_DATE_TIME or the ZonedDateTime." +
                            "String to be parsed: " + e.getParsedString());
                } catch (Exception e) {
                    LOGGER.error("Exception occurred, while PostProcessing the date. Exception message: " + e.getMessage());
                }
            }
```

My issue here is now: If I don't catch the DateTimeParseException, each value of dateFieldName that is not a proper ISO-8601 String would fail the PostProcessor and thus, failing the whole connector. However I don't want the connector to be failing, but I rather want the connector to ignore this event and don't write it into MongoDB.

What I did now to make it work is a bit adventurous:
If the parsing fails in the DateTimeConversionProcessor, I'm calling getNext like this:
```
getNext().ifPresent(pp -> {
                BsonDocument keyDoc = doc.getKeyDoc().isPresent() ? doc.getKeyDoc().get().clone() : null;
                pp.process(new SinkDocument(keyDoc, null), orig);
            });
```

In the connector config, I put the DocumentIdAdder at the end, after the DateTimeConversionProcessor:
```
"mongodb.post.processor.chain": "at.grahsl.kafka.connect.mongodb.processor.DateTimeConversionProcessor,at.grahsl.kafka.connect.mongodb.processor.DocumentIdAdder"
```

Which stops the DocumentIdAdder from writing an "_id" field, if the valueDoc was emptied on the last processor in the chain.

And finally I changed the if condition after the processorChain.process call in MongoDBSinkTask.buildWriteModel(..) to
```
processorChain.process(doc, record);
if(doc.getValueDoc().isPresent() && doc.getValueDoc().get().containsKey("_id")) {
```

So that only documents are added to the docsToWrite List, if they have an _id field.


Long story short: 
**How would you do the ExceptionHandling, in case you don't want to fail the connector on wrong dateTimeField entries?**

**Comment by hpgrahsl** [comment]
Thx @CGaul for the clarification of what you want to achieve. Let me briefly reflect on this and share some of my immediate thoughts:

Like you already said it's a bit adventurous to do it like that. At the same time I have to admit there are'nt particularly much better ways to achieve what you want to have. The design of the current PostProcessor capability hasn't been done with such use cases in mind thus it's not really "prepared" for this.

On the other hand the principal idea  - at least to my humble understanding - behind Kafka Connect API is to effectively always maintain a 1:1 relation between a source record and a sink record. That being said, it's most likely not the right approach to ignore errors in order to drop records. For several use cases this would even be harmful since it will almost certainly lead to "inconsistent state" at the sink side.

In your case you should actually ask yourself why it's even possible/allowed in the first place to end up with syntactically/semantically invalid records that are written to the kafka topic in question. Ideally you have ways and means in place that such naiive issues like formatting errors don't find their way into persisted records. If for whatever reason that's not an option since you're maybe not in charge of the producing side of the data flow, then the next best thing in my opinion would be to let a kstreams app run over the topic, clean the records up or filter out inconsistent ones and only write the valid ones to a new topic which will be processed by means of a sink connector.

Happy to hear your thoughts!

Cheers, Hans-Peter

**Comment by CGaul** [comment]
Hey Hans-Peter,

sorry for the late reply, this week was rather busy. 
I think the principal idea of maintaining a 1:1 relation in a Kafka-Connector makes sense, so that Kafka-Connect itself can be seen as a lightweight framework for record passing only. 
Just my two cents to add here: The chain of responsibility pattern also defines the notion of a "dispatcher". This could be used to maintain the 1:1 relation, where valid BsonDocuments go to MongoDB-Table 1 and invaid BsonDocuments go to MongoDB-Table2. However I also agree that this kind of validation and dispatching is better off in a kafka streams application (where valid records go to kafka-sink-topic1 and invalid ones to kafka-sink-topic2), so that the connector logic is not going to be overengineered.

In my usecase, I unfortunately have no influence on the producing side and the avro schema is rather "open" (e.g. timestamp is a String), so that invalid timestamps can make it to the topic in question (let's call it source topic). That's why I have to do a validation on the source topic, before writing records to MongoDB.

In summary I agree that it will make more sense to do a validation and dispatching with Kafka-Streams and in a second step have a mongoDB connector only on the validated topic, were records are written to by the kafka-streams application.

Thanks for the very helpful discussion.
Sincerely,
Constantin.

P.S. I will close this "Issue" now. :)

**Comment by hpgrahsl** [comment]
@CGaul this could be of interest for you. seems like a future version of Kafka Connect will get a feature which at least will give you basic support for your potential use case. see here https://cwiki.apache.org/confluence/display/KAFKA/KIP-298%3A+Error+Handling+in+Connect
It's very helpful since if this gets available more or less all available connectors (including my mongodb-sink-connector) will directly benefit from it.

**Comment by CGaul** [comment]
Hey @hpgrahsl Thanks a lot! That indeeds is an interesting article, related to my original problem.
